portalnet/src/overlay_service.rs

```diff
@@ -1727,67 +1794,63 @@ where
         }
     }
 
-    fn process_received_content(&mut self, content: Vec<u8>, request: FindContent) {
-        let content_key = match TContentKey::try_from(request.content_key) {
-            Ok(val) => val,
-            Err(msg) => {
-                error!(
-                    protocol = %self.protocol,
-                    error = ?msg,
-                    "Error decoding content key requested by local node"
-                );
-                return;
-            }
-        };
-        let content_id = content_key.content_id();
-
-        match self
-            .store
-            .read()
-            .is_key_within_radius_and_unavailable(&content_key)
-        {
-            Ok(true) => {
-                let validator = Arc::clone(&self.validator);
-                let store = Arc::clone(&self.store);
-                // Spawn task that validates content before storing.
-                // Allows for non-blocking requests to this/other overlay services.
-                tokio::spawn(async move {
-                    if let Err(err) = validator.validate_content(&content_key, &content).await {
-                        warn!(
-                            error = ?err,
-                            content.id = %hex_encode_compact(content_id),
-                            content.key = %content_key,
-                            "Error validating content"
-                        );
-                        return;
-                    };
-
-                    if let Err(err) = store.write().put(content_key.clone(), content) {
-                        error!(
-                            error = %err,
-                            content.id = %hex_encode_compact(content_id),
-                            content.key = %content_key,
-                            "Error storing content"
-                        );
-                    }
-                });
-            }
-            Ok(false) => {
-                debug!(
+    #[allow(clippy::too_many_arguments)]
+    fn process_received_content(
+        kbuckets: Arc<RwLock<KBucketsTable<NodeId, Node>>>,
+        command_tx: UnboundedSender<OverlayCommand<TContentKey>>,
+        validator: Arc<TValidator>,
+        store: Arc<RwLock<TStore>>,
+        content: (Vec<u8>, bool),
+        content_key: TContentKey,
+        responder: Option<oneshot::Sender<FindContentResult>>,
+        trace: Option<QueryTrace>,
+        nodes_to_poke: Vec<NodeId>,
+    ) {
+        // Spawn task that validates content before storing.
+        // Allows for non-blocking requests to this/other overlay services.
+        tokio::spawn(async move {
+            let content_id = content_key.content_id();
+            if let Err(err) = validator.validate_content(&content_key, &content.0).await {
```

Comment:
Previously, the content was only validated after checking if it's in-radius and missing from the DB. That seems like the right choice. Why was it changed to happen before that check?

I guess it's because we might be poking other nodes with it, but I think there are better solutions. Something like:

```
if is_locally_stored:
  # ignore retrieved value (unless we want to punish the peer later, we can validate with `db.get(key) == content`
  content = db.get(key)
elif in_radius:
  validate(content)
  store(content)
poke(content)
```